improve DetailsList row performance

OneCyrus opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #3853 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

added shouldComponentUpdate to detailsrow to only render the row when the props change

#### Focus areas to test

any edge cases where rows should render again when not changing the props

JasonGore commented:
I gave this change a try out of curiosity and I'm not seeing any noticeable differences in behavior. Can you quantify the improvement in any way?

dzearing commented:
@JasonGore This change may reduce the number of renders that occur on scrolls. @OneCyrus it would be good to enumerate exactly the steps you were testing with.

You can probably see it by providing a cell render that renders the "current render count"... today it may redundantly render things at certain times, It's possible this could avoid some re-renders if the parent re-renders and passes the same content into the List.

This particular change runs the risk of potentially avoiding an update that someone was previously relying on. I'm wondering if this should be opt in to avoid a potential break.

JasonGore commented:
@dzearing I think it should be up to the author to quantify the effects seen and/or improvement gain. 😄 

If stores don't mutate state (as is standard in React), is it not enough just to do a reference comparison in cases like these?

dzearing commented:
The screener tests reveal that this breaks selection. We would need to address this to consider this PR.

cliffkoh commented:
So, this means that even within the realm of the component, it doesn't treat the props as immutable. It's very likely that this little change might end up breaking partners. We therefore would need to make this an opt-in optimization to minimize disruption... (in addition to fixing the selection regression caught by screener tests)

aditima commented:
@OneCyrus, as Cliff and David mentioned above we'd require the following changes to consider your PR:
1. Fix selection as pointer out by screener breaks
2. Add this optimization as an opt-in as this relies on props being immutable, and might end up breaking partners who don't follow that assumption

OneCyrus commented:
I fixed the selection issue and added a feature flag. the main issue with the details list is the architecture of the data flow. currently the details row needs two rendering for changing the selection state. a proper refactoring could help this component quite a bit. but with this change there should be a big improvement already.

dzearing commented:
@OneCyrus I agree it absolutely needs a refactoring. I've started looking at the underlying List component first, as that will have the biggest impact on perf improvement.

dzearing commented:
@aditima Are you ok with this change?

cliffkoh commented on packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.types.ts:
```diff
@@ -245,6 +245,11 @@ export interface IDetailsListProps extends React.Props<DetailsList>, IWithViewpo
   enterModalSelectionOnTouch?: boolean;
+
+  /**
+   * Rerender DetailsRow only when props changed. Might cause regression when depending on external updates.
```
Would be good to document the default value via `@default`. This also makes it into the autodoc.

cliffkoh commented on common/changes/office-ui-fabric-react/fix_detailslist_2018-05-13-19-01.json:
```diff
@@ -0,0 +1,11 @@
+    {
+      "packageName": "office-ui-fabric-react",
+      "comment": "DetailsList: DetailsRow now behaves as a pure component, resulting in improved performance.",
+      "type": "patch"
```
Because of the new prop, no longer a patch change...

aditima commented:
@OneCyrus - could you please sync and merge with master?

dzearing commented:
@OneCyrus any update?

OneCyrus commented:
i need to fix the merge for version 6. i'll have a look at the weekend.

OneCyrus commented:
needed to create a new PR #5571. I'll close this one.
